### PR TITLE
Add proportional mob stack harvest drops and persist potion effects over mob stack kills

### PR DESCRIPTION
**Pull request by broman**
Implements #2 

I think I did this right. Literally of this code was produced between 12am and 2am and I took a page out of Sam's book and didn't test anything, so don't yell at me

### CODE REVIEW

File: src/main/scala/com/mewna/mc/box/components/mobstacking/MobHarvestListener.scala [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      if(item.getType == Material.SHEARS) {
+        val le = entity.asInstanceOf[LivingEntity]
+        for(_ <- 2 until component.count(le))
+          // We don't apply damage on the first shear since the game takes care of that.
+          applyDamage(item, 1)
+      }
+    }
+  }
+  //noinspection ScalaStyle
+  def applyDamage[T<: ItemStack with Damageable](item: T, damage: Int): Unit = {
```
Comment (queer) [inline src/main/scala/com/mewna/mc/box/components/mobstacking/MobHarvestListener.scala]: `T<:` should be `T <:` just to look the same as the rest of the code

File: src/main/scala/com/mewna/mc/box/components/mobstacking/MobHarvestListener.scala [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    // Applies damage to a given ItemStack, factoring in the Unbreaking enchantment
+    val unbreakingLevel = item.getEnchantmentLevel(Enchantment.DURABILITY)
+    val r = new Random()
+    // Formula taken from https://minecraft.gamepedia.com/Unbreaking
+    val unbreakingCalculation = (100 / (unbreakingLevel + 1))
```
Comment (queer) [inline src/main/scala/com/mewna/mc/box/components/mobstacking/MobHarvestListener.scala]: Don't need outer parens here

File: src/main/scala/com/mewna/mc/box/components/mobstacking/MobHarvestListener.scala [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    // TODO Add checking for armour to generalise this method
```
Comment (queer) [inline src/main/scala/com/mewna/mc/box/components/mobstacking/MobHarvestListener.scala]: Unbreaking only matters for the item its on, not applied over the whole inventory. Unsure what you mean here exactly.

File: src/main/scala/com/mewna/mc/box/components/mobstacking/MobHarvestListener.scala [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+  @EventHandler
+  def onPlayerInteractWithEntity(event: PlayerInteractAtEntityEvent): Unit = {
+    // Apply proportionate damage to shears when harvesting a sheep stack
+    val entity = event.getRightClicked
+    if(entity.getType == EntityType.SHEEP) {
+      val inventory = event.getPlayer.getInventory
+      val item = if (event.getHand == EquipmentSlot.HAND) {
+        inventory.getItemInMainHand
+      } else { inventory.getItemInOffHand }
```
Comment (queer) [inline src/main/scala/com/mewna/mc/box/components/mobstacking/MobHarvestListener.scala]: Shouldn't be all in one line

File: src/main/scala/com/mewna/mc/box/components/mobstacking/MobHarvestListener.scala [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  }
+  //noinspection ScalaStyle
+  def applyDamage[T<: ItemStack with Damageable](item: T, damage: Int): Unit = {
+    // Applies damage to a given ItemStack, factoring in the Unbreaking enchantment
+    val unbreakingLevel = item.getEnchantmentLevel(Enchantment.DURABILITY)
+    val r = new Random()
+    // Formula taken from https://minecraft.gamepedia.com/Unbreaking
+    val unbreakingCalculation = (100 / (unbreakingLevel + 1))
+    // TODO Add checking for armour to generalise this method
+    if(r.nextInt(100) < unbreakingCalculation) { // "magic number reeee"
```
Comment (queer) [inline src/main/scala/com/mewna/mc/box/components/mobstacking/MobHarvestListener.scala]: Can just suppress the inspection for this line

**Review by queer — CHANGES_REQUESTED** [review]

**Comment by broman** [comment]
Natan pointed out the for loop should be 2 to component.count(le) so I will correct this as well as the reviews above when I get home

File: src/main/scala/com/mewna/mc/box/components/mobstacking/MobHarvestListener.scala (hunk 3 above)
Comment (broman) [reply src/main/scala/com/mewna/mc/box/components/mobstacking/MobHarvestListener.scala]: In this case the ItemStack passed to applyDamage is a “stack” of one. getItemInMainHand() returns an ItemStack. In this case getEnchantmentLevel is called on the item used to interact with the sheep

**Comment by broman** [comment]
Requested changes have been implemented

**Review by broman — COMMENTED** [review]
Done!

**Comment by queer** [comment]
@natanbc summoning you in here to make sure I'm not missing anything obvious

@broman Did you actually test it?

**Review by queer — APPROVED** [review]

**Comment by broman** [comment]
> @broman Did you actually test it?

Haha yes
Am testing and currently fixing *minor* issues

**Comment by broman** [comment]
Since I can't work git properly this PR also implements #1

File: src/main/scala/com/mewna/mc/box/components/mobstacking/MobKillListener.scala [hunk 6]
```diff
@@ -17,11 +20,13 @@ class MobKillListener extends Listener {
   @EventHandler
   def onKill(event: EntityDeathEvent): Unit = {
     val entity = event.getEntity
+    val effects: util.Collection[PotionEffect] = entity.getActivePotionEffects
     if(component.isStackable(entity)) {
       val count = component.count(entity)
       if(count > 1) {
         val next = count - 1
         val clone = entity.getWorld.spawnEntity(entity.getLocation, entity.getType).asInstanceOf[LivingEntity]
+        clone.addPotionEffects(effects)
```
Comment (queer) [inline src/main/scala/com/mewna/mc/box/components/mobstacking/MobKillListener.scala:29]: I assume you tested this?

Comment (broman) [reply src/main/scala/com/mewna/mc/box/components/mobstacking/MobKillListener.scala:29]: This time yes!

Comment (broman) [reply src/main/scala/com/mewna/mc/box/components/mobstacking/MobKillListener.scala:29]: This time yes!

**Pull request merged**